### PR TITLE
feat: implement semantic errors

### LINKED ISSUE #8646: Semantic error messages
I'd like for our error handling to be more robust and thinking of ways in which we can control:
- Errors which mean we should abort the whole run (e.g. disk space, authorization)
- Errors which mean we should abort the repo (config error, git changed, etc)
- Errors which mean we should abort the branch (e.g. PR we thought is open is actually closed)

One possibility would be that instead of throwing `new Error(X)` we could have classes of errors e.g. mapped to the above. Something like this could mean we have less string-checking of error messages in our error handling and instead can know whether to pass it up or log and consume it.

Another thing I need from the app perspective is to know which repo-level errors are "permanent" - e.g. disabled by config, a fork, etc.

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

Draft version of semantic errors

## Context:

Closes #8646

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/types/semantic-errors/BranchError.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export class BranchError extends Error {
```
Comment (rarkins) [inline lib/types/semantic-errors/BranchError.ts:1]: Could this just be `types/errors/...` to be shorter?

**Comment by rarkins** [comment]
Looks clean so far. With this current approach would it be "backwards compatible" already? i.e. downstream consumers of the errors will continue doing `if (err.message === 'foo')` and not needing to know it's a `RepositoryError`?

File: lib/types/semantic-errors/BranchError.ts (hunk 1 above)
Comment (viceice) [reply lib/types/semantic-errors/BranchError.ts:1]: I think we should move this to `lib/util/errors` as those are used classes and not type only.

`lib/types` was planned for typescript only types, with are removed on runtime.

**Comment by pret-a-porter** [comment]
> Looks clean so far. With this current approach would it be "backwards compatible" already? i.e. downstream consumers of the errors will continue doing `if (err.message === 'foo')` and not needing to know it's a `RepositoryError`?

Yes, indeed. Each custom error class will extend `Error` class, base class have `message` field, therefore downstream consumers still can detect error by `message`

**Pull request closed without merging**